File: examples/go/go-multi-rpc/lib/cfg/cfg.go
package cfg

import (
	"fmt"
	"os"
	"strings"

	"github.com/google/uuid"

	"github.com/golemcloud/golem-go/binding"
	"github.com/golemcloud/golem-go/golemhost"
)

func ComponentIDFromEnv(key string) (golemhost.ComponentID, error) {
	value := os.Getenv(key)
	if value == "" {
		return [16]byte{}, fmt.Errorf("missing environment variable for component id: %s", key)
	}
	componentID, err := uuid.Parse(strings.ToLower(value))
	if err != nil {
		return [16]byte{}, fmt.Errorf("component id parse failed for %s=%s, %w", key, value, err)
	}
	return golemhost.ComponentID(componentID), nil
}

func ComponentOneID() (golemhost.ComponentID, error) {
	return ComponentIDFromEnv("COMPONENT_ONE_ID")
}

func ComponentTwoID() (golemhost.ComponentID, error) {
	return ComponentIDFromEnv("COMPONENT_TWO_ID")
}

func ComponentThreeID() (golemhost.ComponentID, error) {
	return ComponentIDFromEnv("COMPONENT_THREE_ID")
}

func ComponentOneWorkerURI(workerName string) (binding.GolemRpc0_1_0_TypesUri, error) {
	uri, err := workerURIF(ComponentOneID, workerName)
	return uri, err
}

func ComponentTwoWorkerURI(workerName string) (binding.GolemRpc0_1_0_TypesUri, error) {
	uri, err := workerURIF(ComponentTwoID, workerName)
	return uri, err
}

func ComponentThreeWorkerURI(workerName string) (binding.GolemRpc0_1_0_TypesUri, error) {
	uri, err := workerURIF(ComponentThreeID, workerName)
	return uri, err
}

func WorkerURI[T binding.GolemRpc0_1_0_TypesUri](workerID golemhost.WorkerID) T {
	return T(binding.GolemRpc0_1_0_TypesUri{
		Value: fmt.Sprintf("urn:worker:%s/%s", (uuid.UUID(workerID.ComponentID)).URN(), workerID.WorkerName),
	})
}

func workerURIF(getComponentID func() (golemhost.ComponentID, error), workerName string) (binding.GolemRpc0_1_0_TypesUri, error) {
	componentID, err := getComponentID()
	if err != nil {
		return binding.GolemRpc0_1_0_TypesUri{}, err
	}
	return WorkerURI(golemhost.WorkerID{
		ComponentID: componentID,
		WorkerName:  workerName,
	}), nil
}

File: examples/ts/ts-multi-rpc/src/lib/cfg.ts
import { getEnvironment } from "wasi:cli/environment@0.2.0";

let env: Map<string, string> | undefined = undefined;

export const envVarKeys = {
  COMPONENT_ONE_ID: "COMPONENT_ONE_ID",
  COMPONENT_TWO_ID: "COMPONENT_TWO_ID",
  COMPONENT_THREE_ID: "COMPONENT_THREE_ID",
};

function getEnv(key: string): string | undefined {
  if (env === undefined) {
    env = new Map();
    for (const [key, value] of getEnvironment()) {
      env.set(key, value);
    }
  }

  return env.get(key);
}

function mustGetEnv(key: string): string {
  const value = getEnv(key);
  if (value == undefined) {
    throw new Error(`Expected environment variable is missing: ${key}`);
  }
  return value;
}

export interface Uri {
  value: string;
}

function getComponentWorkerURN(componentID: string, workerName: string): string {
  return `urn:worker:${componentID}/${workerName}`;
}

export function getComponentOneWorkerURN(workerName: string): Uri {
  return {
    value: getComponentWorkerURN(mustGetEnv(envVarKeys.COMPONENT_ONE_ID), workerName),
  };
}

export function getComponentTwoWorkerURN(workerName: string): Uri {
  return {
    value: getComponentWorkerURN(mustGetEnv(envVarKeys.COMPONENT_TWO_ID), workerName),
  };
}

export function getComponentThreeWorkerURN(workerName: string): Uri {
  return {
    value: getComponentWorkerURN(mustGetEnv(envVarKeys.COMPONENT_THREE_ID), workerName),
  };
}
